code/projects/app/src/app/shared/image-comparison/image-comparison.service.ts

```diff
@@ -0,0 +1,49 @@
+import { Injectable } from "@angular/core";
+import { BehaviorSubject, combineLatest, filter, Observable } from "rxjs";
+
+@Injectable({
+  providedIn: "root",
+})
+export class ImageComparisonService {
+  protected left = new BehaviorSubject<ImageData | null>(null);
+  protected right = new BehaviorSubject<ImageData | null>(null);
+
+  public left$ = this.left.pipe(filter((image) => image !== null));
+  public right$ = this.right.pipe(filter((image) => image !== null));
+
+  // typescript please stop complaining
+  public lastImages$ = combineLatest([this.left, this.right]).pipe(
+    filter(([l, r]) => l !== null && r !== null),
+  ) as Observable<[ImageData, ImageData]>;
+
+  public addImage(which: "left" | "right", image: ImageData): Promise<boolean> {
+    // Resolve false only when the image is not the same size as the other image
+    const promise = new Promise<boolean>((resolve) => {
+      const left = this.left.getValue();
+      const right = this.right.getValue();
+      let canAdd = false;
+
+      switch (which) {
+        case "left":
+          canAdd =
+            right == null ||
+            (right.width == image.width && right.height == image.height);
+          if (canAdd) this.left.next(image);
+          break;
+
+        case "right":
+          canAdd =
+            left == null ||
+            (left.width == image.width && left.height == image.height);
+          if (canAdd) this.right.next(image);
+          break;
+      }
+      resolve(canAdd);
```

Comment:
It will be probably removed during the rewrite, but maybe something similar will be handy in the future.

```suggestion
      const values = { left: this.left.getValue(), right: this.right.getValue() };
      const canAdd =
            values[which] === null ||
            (values[which].width === image.width && values[which].height == image.height);
      if (canAdd) this[which].next(image);
      resolve(canAdd);
```

Reply:
I see that you want to optimize-out that `switch` statement, but I don't think we can do that. Look at `case` name and then look at the boolean statement. Those check opposing values... If we want to add any image, we have to check whether the ***other one*** is the same size as the one we are trying to add.